rust/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+/*
+ * To the extent possible under law, the author has dedicated all copyright
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * See <http://creativecommons.org/publicdomain/zero/1.0/>
+ */
+
+#![feature(pattern)]
+
+use std::cmp::{
+    Ordering,
+    Ordering::{Equal, Greater, Less},
+};
+use std::collections::VecDeque;
+use std::str::pattern::Pattern;
+
+#[derive(Debug)]
+enum SortingType {
+    Numerical(i64),
+    Lexical(String),
+    SemverPrerelease(String),
+}
+
+impl SortingType {
+    fn to_string(self) -> String {
+        match self {
+            Self::Numerical(a) => a.to_string(),
+            Self::Lexical(a) | Self::SemverPrerelease(a) => a,
+        }
+    }
+}
+
+fn split_once_rest<'a, P>(s: &'a str, pat: P) -> Option<(&str, &str)>
+where
+    P: Pattern<'a>,
+{
+    let loc = s.find(pat);
+    if let Some(index) = loc {
+        Some(s.split_at(index))
+    } else {
+        Some((s, &""))
+    }
+}
+
+fn is_semver_prerelease(s: &str) -> bool {
+    s.len() > 1 && s.starts_with('-')
+}
+
+fn decompose(str_in: &str) -> VecDeque<SortingType> {
+    if str_in.is_empty() {
+        return VecDeque::new();
+    }
+
+    let mut last_numeric = str_in.starts_with(char::is_numeric);
+    let mut s = str_in.to_owned();
+    let mut out: VecDeque<SortingType> = VecDeque::new();
+
+    if let Some((left, _)) = s.split_once('+') {
+        s = left.to_owned();
+    };
+
+    while !s.is_empty() {
+        if last_numeric {
+            if let Some((left, right)) = split_once_rest(&s, |c: char| !c.is_numeric()) {
```

Comment:
`is_numeric` is defined to return `true` if the Unicode codepoint is any numerical character class; this won't necessarily ensure the string is parseable as an `i64` and may cause the program to panic. Java's `isDigit` matches only character ranges that Java knows how to parse; unfortunately, Rust's equivalent `is_digit` only matches ASCII digits...

It's probably acceptable to only allow ASCII numbers, and I may want to modify the Java implementation in the same way.

Reply:
Definitely a good point. Restricting to ASCII digits would probably be best. Aligns with SemVer, and reduces surprises.